nixos/modules/system/activation/switch-to-configuration.py

```diff
@@ -0,0 +1,604 @@
+#! @python@
+
+import os.path
+import subprocess
+from syslog import openlog, syslog, LOG_NOTICE, LOG_ERR
+import sys
+
+out = "@out@"
+
+# To be robust against interruption, record what units need to be started etc.
+startListFile = "/run/systemd/start-list"
+restartListFile = "/run/systemd/restart-list"
+reloadListFile = "/run/systemd/reload-list"
+
+def die_with_usage():
+    print("""
+        Usage: {prog} [switch|boot|test]
+
+        switch:       make the configuration the boot default and activate now
+        boot:         make the configuration the boot default
+        test:         activate the configuration, but don\'t make it the boot default
+        dry-activate: show what would be done if this configuration were activated
+    """.format(prog=sys.argv[0]), file=sys.stderr)
+    sys.exit(1)
+
+def die(why):
+    print("error:", why, file=sys.stderr)
+    sys.exit(1)
+
+def read_if_exists(path, otherwise=None):
+    if os.path.isfile(path):
+        with open(path, 'r') as f:
+            return f.read()
+    else:
+        return otherwise
+
+if len(sys.argv) < 1:
+    die_with_usage()
+
+action = sys.argv[1]
+valid_modes = ["switch", "boot", "test", "dry-activate"]
+if action not in valid_modes:
+    die_with_usage()
+
+# This is a NixOS installation if it has /etc/NIXOS or a proper
+# /etc/os-release.
+def is_nixos():
+    if os.path.isfile("/etc/NIXOS"):
+        return True
+
+    os_release = read_if_exists("/etc/os-release", otherwise="")
+    if os_release.find("ID=nixos") != -1:
+        return True
+
+    return False
+
+if not is_nixos():
+    die("This is not a NixOS installation!")
+
+openlog("nixos")
+
+# Install or update the bootloader.
+if action == "switch" or action == "boot":
+    subprocess.checkCall(["@installBootLoader@", out])
+
+# Just in case the new configuration hangs the system, do a sync now.
+if os.environ.get('NIXOS_NO_SYNC') != "1":
+    subprocess.checkCall(["@coreutils@/bin/sync", "-f", "/nix/store"])
+
+if action == "boot":
+    sys.exit(0)
+
+# Check if we can activate the new configuration.
+oldVersion = read_if_exists("/run/current-system/init-interface-version", otherwise="")
+newVersion = open(os.path.join(out, "init-interface-version")).read()
+
+if newVersion != oldVersion:
+    print("""
+        Warning: the new NixOS configuration has an ‘init’ that is
+        incompatible with the current configuration.  The new configuration
+        won\'t take effect until you reboot the system.
+    """, file=sys.stderr)
+    sys.exit(100)
+
+# Ignore SIGHUP so that we're not killed if we're running on (say)
+# virtual console 1 and we restart the "tty1" unit.
+import signal
+signal.signal(signal.SIGPIPE, signal.SIG_IGN)
+
+def get_active_units():
+    import dbus
```

Comment:
move import to top

Reply:
I was somewhat intentionally leaving this here to allow for the possibility of disabling the dbus logic for ease of cross-compilation. However, `dbus-python` is pretty easy to cross so it seems this was unnecessary.